lib/iris/aux_factory.py

```diff
@@ -784,20 +792,65 @@ def dependencies(self):
                     depth_c=self.depth_c, nsigma=self.nsigma, zlev=self.zlev)
 
     def _derive(self, sigma, eta, depth, depth_c,
-                zlev, shape, nsigma_slice):
+                zlev, nsigma, coord_dims_func):
+        # Calculate the index of the 'z' dimension in the input arrays.
+        # First find the cube 'z' dimension ...
+        [cube_z_dim] = coord_dims_func(self.dependencies['zlev'])
+        # ... then calculate the corresponding dependency dimension.
+        derived_cubedims = self.derived_dims(coord_dims_func)
+        z_dim = derived_cubedims.index(cube_z_dim)
+
+        # Calculate the result shape as a combination of all the inputs.
+        # Note: all the inputs have the same number of dimensions >= 1, except
+        # for any missing dependencies, which have scalar values.
+        allshapes = np.array(
+            [el.shape
+             for el in (sigma, eta, depth, depth_c, zlev)
+             if el.ndim > 0])
+        result_shape = list(np.max(allshapes, axis=0))
+        ndims = len(result_shape)
+
+        # Make a slice tuple to index the first nsigma z-levels.
+        z_slices_nsigma = [slice(None)] * ndims
+        z_slices_nsigma[z_dim] = slice(0, int(nsigma))
+        z_slices_nsigma = tuple(z_slices_nsigma)
+        # Make a slice tuple to index the remaining z-levels.
+        z_slices_rest = [slice(None)] * ndims
+        z_slices_rest[z_dim] = slice(int(nsigma), None)
+        z_slices_rest = tuple(z_slices_rest)
```

Comment:
The `z_slices_nsigma` thing is a tuple of keys to extract the first 'nsigma' z-levels, 
i.e. `data[z_slices_nsigma]` would be something line `data[:, ..., :, :nsigma, :, ..., :]`
 \- recall that here, 'nsigma' is *just a number*.

This is exactly what in the original code is called `nsigma_slice`.
I changed over to calculating that *within* the `_derive` call, instead of passing in in from the `make_coord` routine, as I thought it was much clearer to have derivation + use of it in the same place.

Meanwhile, the subsequent bit `z_slices_rest` is to select the "remaining" z levels, i.e. those beyond the first nsigma : 
that is, if `data[z_slices_nsigma]` does something like `data[:, ..., :, `**`0:nsigma`**`, :, ..., :]`, 
then `data[z_slices_rest]` is like `data[:, ..., :, `**`nsigma:-1`**`, :, ..., :]`, 
In the new calc, we need to extract over the 'remaining' levels to get the concatenate right.

Is that any clearer ?